graph/db/sql_store.go

```diff
@@ -80,3 +131,749 @@ func NewSQLStore(db BatchedSQLQueries, kvStore *KVStore,
+	}
+
+	return nil
+}
+
+// FetchNodeFeatures returns the features of the given node. If no features are
+// known for the node, an empty feature vector is returned.
+//
+// NOTE: this is part of the graphdb.NodeTraverser interface.
+func (s *SQLStore) FetchNodeFeatures(nodePub route.Vertex) (
+	*lnwire.FeatureVector, error) {
+
+	ctx := context.TODO()
+
+	return fetchNodeFeatures(ctx, s.db, nodePub)
+}
+
+// getNodeByPubKey attempts to look up a target node by its public key.
+func getNodeByPubKey(ctx context.Context, db SQLQueries,
+	pubKey route.Vertex) (int64, *models.LightningNode, error) {
+
+	dbNode, err := db.GetNodeByPubKey(
+		ctx, sqlc.GetNodeByPubKeyParams{
+			Version: int16(ProtocolV1),
+			PubKey:  pubKey[:],
+		},
+	)
+	if errors.Is(err, sql.ErrNoRows) {
+		return 0, nil, ErrGraphNodeNotFound
+	} else if err != nil {
+		return 0, nil, fmt.Errorf("unable to fetch node: %w", err)
+	}
+
+	node, err := buildNode(ctx, db, &dbNode)
+	if err != nil {
+		return 0, nil, fmt.Errorf("unable to build node: %w", err)
+	}
+
+	return dbNode.ID, node, nil
+}
+
+// buildNode constructs a LightningNode instance from the given database node
+// record. The node's features, addresses and extra signed fields are also
+// fetched from the database and set on the node.
+func buildNode(ctx context.Context, db SQLQueries, dbNode *sqlc.Node) (
+	*models.LightningNode, error) {
+
+	if dbNode.Version != int16(ProtocolV1) {
+		return nil, fmt.Errorf("unsupported node version: %d",
+			dbNode.Version)
+	}
+
+	var pub [33]byte
+	copy(pub[:], dbNode.PubKey)
+
+	node := &models.LightningNode{
+		PubKeyBytes:     pub,
+		Features:        lnwire.EmptyFeatureVector(),
+		LastUpdate:      time.Unix(0, 0),
+		ExtraOpaqueData: make([]byte, 0),
+	}
+
+	if len(dbNode.Signature) == 0 {
+		return node, nil
+	}
+
+	node.HaveNodeAnnouncement = true
+	node.AuthSigBytes = dbNode.Signature
+	node.Alias = dbNode.Alias.String
+	node.LastUpdate = time.Unix(dbNode.LastUpdate.Int64, 0)
+
+	var err error
+	node.Color, err = DecodeHexColor(dbNode.Color.String)
+	if err != nil {
+		return nil, fmt.Errorf("unable to decode color: %w", err)
+	}
+
+	// Fetch the node's features.
+	node.Features, err = getNodeFeatures(ctx, db, dbNode.ID)
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch node(%d) "+
+			"features: %w", dbNode.ID, err)
+	}
+
+	// Fetch the node's addresses.
+	_, node.Addresses, err = getNodeAddresses(ctx, db, pub[:])
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch node(%d) "+
+			"addresses: %w", dbNode.ID, err)
+	}
+
+	// Fetch the node's extra signed fields.
+	extraTLVMap, err := getNodeExtraSignedFields(ctx, db, dbNode.ID)
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch node(%d) "+
+			"extra signed fields: %w", dbNode.ID, err)
+	}
+
+	recs, err := lnwire.CustomRecords(extraTLVMap).Serialize()
+	if err != nil {
+		return nil, fmt.Errorf("unable to serialize extra signed "+
+			"fields: %w", err)
+	}
+
+	if recs != nil {
```

Comment:
nit: length check instead to avoid adding an empty map? Or is this on purpose for easier comparison in unit tests?